### PR TITLE
Add support for top_p in non-deterministic generation

### PR DESCRIPTION
**Pull request by Adibvafa**
This PR adds support for top_p and more test cases for predict_dna_sequence.

### CODE REVIEW

**Review by gui11aume — COMMENTED** [review]
Excellent work! I have some optional comments. The only serious one is about removing the for loop in a place where may need performance.

File: tests/test_CodonPrediction.py [hunk 1]
```diff
@@ -71,7 +78,7 @@ def test_predict_dna_sequence_non_deterministic(self):
 
     def test_predict_dna_sequence_invalid_inputs(self):
         test_cases = [
-            ("MKTZZFVLLL", "Escherichia coli general", "invalid protein sequence"),
+            ("MKTZZFVLLL?", "Escherichia coli general", "invalid protein sequence"),
```
[the remark this answers is not in the document]
Comment (Adibvafa) [reply tests/test_CodonPrediction.py:81]: Current code does not expect throwing an error when passing ambiguous aminoacids as they are siletnly converted. We will deal with this in thttps://github.com/Adibvafa/CodonTransformer/pull/9

File: CodonTransformer/CodonPrediction.py [hunk 2]
```diff
@@ -210,6 +207,75 @@ def predict_dna_sequence(
...
+            sorted_probs = sorted_probs[: cutoff_index + 1]
+            sorted_indices = sorted_indices[: cutoff_index + 1]
+
+        # Re-normalize the probabilities after filtering
+        filtered_probs = sorted_probs / sorted_probs.sum()
+
+        # Sample from the filtered distribution
+        sampled_index = torch.multinomial(filtered_probs, num_samples=1).item()
+        predicted_index = sorted_indices[sampled_index].item()
+        predicted_indices.append(predicted_index)
```
[the remark this answers is not in the document]
Comment (Adibvafa) [reply CodonTransformer/CodonPrediction.py:274]: This is a great suggestion, I will fix it.

[the remark this answers is not in the document]
Comment (Adibvafa) [reply CodonTransformer/CodonPrediction.py:274]: I ended up using Hugging Face's way!
```
def sample_top_p(probs, p):
    """
    Perform top-p (nucleus) sampling on a probability distribution.

    Args:
        probs (torch.Tensor): Probability distribution tensor.
        p (float): Probability threshold for top-p sampling.

    Returns:
        torch.Tensor: Sampled token indices.

    Note:
        Top-p sampling selects the smallest set of tokens whose cumulative probability mass
        exceeds the threshold p. The distribution is renormalized based on the selected tokens.

    """
    probs_sort, probs_idx = torch.sort(probs, dim=-1, descending=True)
    probs_sum = torch.cumsum(probs_sort, dim=-1)
    mask = probs_sum - probs_sort > p
    probs_sort[mask] = 0.0
    probs_sort.div_(probs_sort.sum(dim=-1, keepdim=True))
    next_token = torch.multinomial(probs_sort, num_samples=1)
    next_token = torch.gather(probs_idx, -1, next_token)
    return next_token

**Comment by Adibvafa** [comment]
@gui11aume Thank you for the comments, they were quite helpful. The PR is merged.

**Pull request merged**